### PR TITLE
Restore flups part 2: Retry if metadata not found

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**:

==COMMIT_MSG==
The Atlas restore service will now retry if metadata is not found for a given table.
==COMMIT_MSG==

**Implementation Description (bullets)**:
- Moved session creation retry logic into CqlSession
- Introduced a new method `CqlSession.getTableMetadata`, with new retry logic, and ensured we always use this to get table metadata.

**Testing (What was existing testing like?  What have you done to improve it?)**:
TODO - add tests about the metadata updates

**Concerns (what feedback would you like?)**: The session refresh logic used internally is quite involved and would need a more substantial refactor. Considering that, is it worthwhile to close and reopen the session?

**Where should we start reviewing?**: CqlSession

**Priority (whenever / two weeks / yesterday)**: today

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/TokenRangeFetcher.java [hunk 1]
```diff
@@ -58,22 +56,7 @@ public TokenRangeFetcher(
...
-
-        TableMetadata tableMetadata = keyspaceMetadata.getTable(tableName);
-        if (tableMetadata == null) {
-            log.error(
-                    "Could not find metadata for table that is supposed to exist",
-                    SafeArg.of("keyspace", namespace),
-                    SafeArg.of("tableName", tableName));
-            return ImmutableMap.of();
-        }
+        TableMetadata tableMetadata = cqlSession.getTableMetadata(tableName);
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/TokenRangeFetcher.java:59]: do you need to add retrying logic to any other members of the class?

**Review by sudiksha27 — COMMENTED** [review]
only looked at correctness and not at refactor itself or styling. I think this looks fine - once we have confirmed the refresh logic with @zpear i can take do a second pass.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/TokenRangeFetcher.java (hunk 1 above)
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/TokenRangeFetcher.java:59]: all the table metadata fetching logic should be fed through CqlSession now. We haven't seen the other calls failing.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlSession.java [hunk 2]
```diff
@@ -45,6 +110,54 @@ public CqlMetadata getMetadata() {
...
+            throw new SafeIllegalStateException(
+                    "Failed to execute CqlSession connect even with retry",
+                    e,
+                    SafeArg.of("namespace", namespace),
+                    SafeArg.of("tableName", tableName));
+        }
+    }
+
+    private void refreshSession() {
+        session.closeAsync();
```
Comment (gmaretic) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/CqlSession.java:148]: Do we want to force here?

**Review by gmaretic — COMMENTED** [review]
I think this makes sense overall -- do we want to cut an rc and verify it helps?

**Pull request closed without merging**